Merge in mainline decaf fixes

GoodForOneFare opened the pull request:
Merges in recent fixes from `juliankrispel/decaf`.  This should get us closer to submitting stuff back to mainline, and not maintaining our own fork.

Also includes 3 tweaks:
- For inline assignments, generate ternary assignments (mainline now prefers IIFEs)
- Tweak to `shopify/decaf` logic to prevent duplicate local var declarations
- Tweak to `shopify/decaf` logic to declare bulk-destructuring-assignment vars at the "top" of scope

/cc @lemonmade, @fandy

lemonmade commented:
@GoodForOneFare what do you think is the best way to review this?

GoodForOneFare commented:
I'll add some comments to the mainline merge (286ffc5) pointing out where I've fixed conflicts.  The other 3 commits should be reasonably sane taken in isolation.

Anything that I don't comment on can be assumed to be mainline (it's mostly fixes to improve switch-returns, super calls, and class properties), but feel free to ask if something doesn't look sane.  At worst, my response will be, "mainline did it".

GoodForOneFare commented on src/parser.js:
```diff
@@ -255,12 +254,12 @@ function mapClassBodyElement(node, meta) {
   let elementType = 'method';
   let isStatic = false;
-  const type = node.constructor.name;
+  // const type = node.constructor.name;
```
This commented code came directly from mainline.  I've left it in for future merge-friendliness  ¯_(ツ)_/¯

GoodForOneFare commented on src/parser.js:
```diff
@@ -530,41 +538,27 @@ function mapIfStatement(node, meta) {
-      return false;
-  }
-}
+function isTernaryOperation(node, isExpressionCondition) {
```
Manual change to get rid of IIFEs for assignments ([mainline code](https://github.com/juliankrispel/decaf/blob/master/src/parser.js#L521)).

Mainline was generating stuff like:

``` js
 return $(this).addClass((() => {
   if (!$(this).hasClass('disabled')) {
     return 'is-active';
   }
 })());
```

This tweak results gets us back to:

``` js
  return $(this).addClass((!$(this).hasClass('disabled') ? 'is-active' : undefined));
```

I dislike having to pass around this parameter like this.  Maybe it'd be easier to just keep the IIFEs and transform them out?

lemonmade replied:
Yeah, my vote would be to do this as a transform instead of passing the state around.

GoodForOneFare commented on src/parser.js:
```diff
@@ -1084,26 +1143,29 @@ function insertVariableDeclarations(ast) {
+      if (path.parent && path.parent.value.type === 'ExpressionStatement' && isArrayAssignmentWithThisReference(path)) {
+        insertArrayAssignmentVarDeclarations(path);
+      } else if (path.parent &&
+          get(path, 'parent.parent.value.type') !== 'SwitchCase' &&
```
Merge conflict here ([mainline code](https://github.com/juliankrispel/decaf/blob/master/src/parser.js#L975)).  I made the `!== SwitchCase` logic come after our `isArrayAssignmentWithThisReference` logic.

I'll submit the `isArrayAssignmentWithThisReference` fix up as a PR to mainline.

GoodForOneFare commented on src/parser.js:
```diff
@@ -1390,7 +1453,7 @@ function mapNewExpression(node, meta) {
 
 function conditionalStatementAsExpression(node, meta) {
-  const conditionalStatement = mapConditionalStatement(node, meta);
+  const conditionalStatement = mapConditionalStatement(node, meta, true);
```
Passing in `true` here as part of the `isTernaryOperation` IIFE fix.

GoodForOneFare commented on test/test.js:
```diff
@@ -1735,23 +1804,23 @@ describe('comprehensions', () => {
   eat food`;
     const expected =
-`for (let food of ["toast", "cheese", "wine"]) {
+`for (var food of ["toast", "cheese", "wine"]) {
```
All of these loops now use `var` instead of `let`.  Don't know what changed in mainline to cause this.

GoodForOneFare commented:
@lemonmade let me know what else I can do to make this easier.  I can split this into four separate PRs, or add more comments, or...

lemonmade commented:
What you've done is great 👍 will review today.

lemonmade commented:
Thank you so much for doing this. Code LGTM except the ternary thing, which I think will make a nice transform (and keeps us more in line with the main decaf).

GoodForOneFare commented:
Thanks for reviewing the awful mess!  I've removed the custom IIFE/ternary logic, and added the eval-based escaping fix.

A task to clean up IIFEs already exists (#113).  I'll run through the esify output and update the list of affected areas.

I've also added a couple of PRs to mainline that'll get us even more in sync (juliankrispel/decaf#141, juliankrispel/decaf#142).
